### ISSUE TITLE
dev experience with babel

### ISSUE THREAD

**Issue by pkieltyka** [preconstruct/preconstruct#341]
I've tried so many build systems, and have produced many monorepo and project setups for libraries and I just want to say that I'm super fond of preconstruct. It's thoughtful and precise design sensibilities make it my favorite build system macro framework.

Personally I use Typescript for all of my projects, which has also worked well with preconstruct, in combination of course with the babel config, and rollup under-the-hood. However, a lot of projects in the build use Typescript (tsc) directly or rollup+tsc to compile down to es5 with esnext modules. There is a nice simplicity to this for a Typescript environment, but indeed Babel is still a superior transpiler when targetting different environments. 

I'm curious what the authors prioritize for preconstruct's use -- for library development or for front-end (ui, ie. react/vue/angular) web app development? For webapp builds, I still think webpack is king (with new v5 looking pretty cool), but perhaps rollup is a decent choice for front-end apps too?

I do think that preconstruct is a no brainer for library projects targetting multiple outputs (cjs/esm/umd), in an enjoyable dev experience (preconstruct dev), and good build defaults (via rollup). However, I'd like to share a few things on my own experience as developing a library monorepo with preconstruct.

1. I'm only interested in "modern" library constructions that will give me the best experience when consuming the library by other projects, and eventually shipping to end-user programs. The environments include: a.) as a dependency by npm/yarn with another bundler like rollup/webpack/other b.) node package using the es6 module module directly c.) umd bundle for es5 targets with wide browser support and convenience

2. For purposes of #1, it would be great to have these defaults set for preconstruct so things "just work". However, being able to adjust cjs/esm/umd build systems if necessary (but ideally rare).

3.  Is it possible in preconstruct right now to have different transpiling settings between cjs/esm/umd? I haven't finished my experiments with umd in preconstruct, but hopefully I can target es6 in esm, and then es5 with umd

4. For my typescript project, (and I assume would be the same for any es6 library), I had a lot of pains with babel.config.js in my setup. 

Originally my babel setup looked like:

```
module.exports = {
  presets: [
    '@babel/preset-typescript'
  ],
  plugins: [
    ['@babel/plugin-proposal-class-properties', { loose: true }]
  ]
}
```

Which generally worked fine -- it compiled the source down to very modern es6, perhaps es2020..? The issue however, was that the output was a bit *too* modern, which results in my consuming application to fail to load the cjs/esm dist files. Specifically the error was with `const x = another?.field` the optional chaining feature (https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#optional-chaining). It seems the preset keeps the optional chaining, which causes the the webpack esm loader to fail:

```
ERROR in ./node_modules/@0xsequence/provider/dist/provider.esm.js 2862:16
Module parse failed: Unexpected token (2862:16)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
| 
|     this.openWallet = async (path, state) => {
>       if (state?.login !== true && !this.isLoggedIn()) {
|         throw new Error('login first');
|       }
```

it's true that I could setup babel-loader to load the esm, but now it's another experience pain for a consumer. The optional chaining is the only thing making the esm fail by the consuming library. 

In the end, I've setup the following babel.config.js to make my library easy to consume, while still allowing it to be tree-shakable:

```
module.exports = {
  presets: [
    ['@babel/preset-env', {
      targets: {
        esmodules: true
      },
      exclude: [
        '@babel/plugin-transform-async-to-generator',
        '@babel/plugin-transform-regenerator'
      ]
    }],
    '@babel/preset-typescript'
  ],
  plugins: [
    ['@babel/plugin-proposal-class-properties', { loose: true }]
  ]
}
```

My thought would be to perhaps have some "receipes" in preconstruct docs of babel.config.js 's for different scenarios, which I think there are very few (maybe just 2? one for es6 projects and one for ts projects)

Finally, in a larger monorepo project which could include parts of a library and a webapp (via webpack5), is this something preconstruct could support? and if so, how could you vary the babel/rollup config between the respective packages? maybe those are stretch goals for future versions of preconstruct, or maybe it would fit well in your design?

**Comment by pkieltyka** [comment]
another thought would be, for typescript projects, could I skip babel altogether? it's really just causing problems and could I use rollup+tsc directly as an option with preconstruct?

the @babel/preset-env is not configurable enough, and is very difficult. For instance, in my esm output its including..

```
function _defineProperty(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    });
  } else {
    obj[key] = value;
  }

  return obj;
}

function ownKeys(object, enumerableOnly) {
  var keys = Object.keys(object);

  if (Object.getOwnPropertySymbols) {
    var symbols = Object.getOwnPropertySymbols(object);
    if (enumerableOnly) symbols = symbols.filter(function (sym) {
      return Object.getOwnPropertyDescriptor(object, sym).enumerable;
    });
    keys.push.apply(keys, symbols);
  }

  return keys;
}

function _objectSpread2(target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = arguments[i] != null ? arguments[i] : {};

    if (i % 2) {
      ownKeys(Object(source), true).forEach(function (key) {
        _defineProperty(target, key, source[key]);
      });
    } else if (Object.getOwnPropertyDescriptors) {
      Object.defineProperties(target, Object.getOwnPropertyDescriptors(source));
    } else {
      ownKeys(Object(source)).forEach(function (key) {
        Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
      });
    }
  }

  return target;
}
```

I really don't need this in my esm output, which happens because of the @babel/preset-env. Ideally for my "libraries" I could skip this, but then `@babel/preset-typescript` is not configurable enough -- for instance, it doesn't use my tsconfig.json in my project to respect my target, and it also doesn't let me pass a target, the docs explicitly say to use @babel/preset-env. For libraries built with Typescript which are to be consumed by another bundler, babel is just getting in the way. Perhaps you guys know how to wield it though?

**Comment by mitchellhamilton** [comment]
> I'm curious what the authors prioritize for preconstruct's use -- for library development or for front-end (ui, ie. react/vue/angular) web app development?

Preconstruct is built for library development, not for apps.

> Is it possible in preconstruct right now to have different transpiling settings between cjs/esm/umd?

It is not. The big reason is that it would mean a sizable build time increase(because babel would have to run over the same file multiple times) for something that the vast majority of people would never use.

> Finally, in a larger monorepo project which could include parts of a library and a webapp (via webpack5), is this something preconstruct could support?

Yes, that should work fine.

>  and if so, how could you vary the babel/rollup config between the respective packages? maybe those are stretch goals for future versions of preconstruct, or maybe it would fit well in your design?

Varying babel config between packages will work completely fine using [Babel's overrides option](https://babeljs.io/docs/en/options#overrides), having babel configs per-package or etc.

Not sure what you mean by varying the Rollup config per package? Preconstruct doesn't allow users to directly change the Rollup config?

> another thought would be, for typescript projects, could I skip babel altogether? it's really just causing problems and could I use rollup+tsc directly as an option with preconstruct?

Using tsc directly isn't likely to be supported because [it raises so many questions and problems](https://github.com/preconstruct/preconstruct/issues/311#issuecomment-679808489) and can't do anything (with the exception of const enums and namespace but those features are very rarely used) that Babel can't do

If you really really wanted to use tsc, you could do something like https://github.com/mitchellhamilton/babel-plugin-tsc (though I wouldn't recommend using that package because it probably won't be very performant) that uses Babel's parserOverride to run tsc before Babel and then don't do any transforms with Babel if you really wanted to.

> I really don't need this in my esm output, which happens because of the @babel/preset-env.

I'm guessing you want to use Object.assign? If so, you can add `["@babel/plugin-proposal-object-rest-spread", { useBuiltIns: true, loose: true }]` to your plugins

**Comment by pkieltyka** [comment]
> Preconstruct is built for library development, not for apps.

I thought so too but I wanted to clarify the design goals. I think that would be a good thing to mention in the README.

> > Is it possible in preconstruct right now to have different transpiling settings between cjs/esm/umd?

> It is not. The big reason is that it would mean a sizable build time increase(because babel would have to run over the same file multiple times) for something that the vast majority of people would never use.

Hmm, if preconstruct is for libraries, wouldn't you agree that dist output should be in es6 so that consumers of the library can tree-shake unused parts of the library? If that is so, wouldn't you agree that es6 is also a poor choice for a umd dist as it wouldn't support enough browsers? Therefore, I think preconstruct has some design challenges here. I hear you on the performance reasoning to have a single babel config -- but I think completeness/correctness/experience is far more important to optimize for than for "performance". As well, I'd argue preconstruct has a great performance story since it has the `preconstruct dev` option of linking the packages together for local development without needing any building whatsoever. The dist builds would happen only on dist releases. Therefore, I think not having support for multiple outputs depending on the module format is a big oversight. Alternatively, the recommendation could be to mention this in the README with a big warning, and then make a recommendation for people to use their own rollup config for umd dist's of es5. In this case, would be useful to have "preconstruct: { umdDisable: true }" so that preconstruct doesn't conflict with other bundlers of the package "umd:main" setting. I'd advocate of course first for preconstruct to handle the "best" case of varying outputs by dist, but I'm offering other ideas so that preconstruct is viable for libraries that need to target module outputs and as well umd with wide compat.

As well, preconstruct should suggest the ideal babel config as well for best experiences as a default. I'd also advocate for better / more prescribed Typescript configs as I spent a lot of extra hours making Typescript feel right with the babel output for library consumption. 

Just an FYI, the babel config for Typescript source that I felt was easy to consume and wasn't over-transpiled is:

```
module.exports = {
  presets: [
    ['@babel/preset-env', {
      targets: {
        esmodules: true
      },
      bugfixes: true,
      loose: true,
      exclude: [
        '@babel/plugin-transform-async-to-generator',
        '@babel/plugin-transform-regenerator'
      ]
    }],
    '@babel/preset-typescript'
  ],
  plugins: [
    ['@babel/plugin-proposal-class-properties', { loose: true }]
  ]
}
```

**Comment by mitchellhamilton** [comment]
> wouldn't you agree that dist output should be in es6 so that consumers of the library can tree-shake unused parts of the library?

Not sure why es6 would be related to tree shaking? I wouldn't really agree that people should ship es6 to npm, especially for front-end code because compilation of node_modules with @babel/preset-env doesn't universally happen(I know there are some tools that do it but it's not a huge amount).

> As well, preconstruct should suggest the ideal babel config as well for best experiences as a default.

Not sure how helpful recommending a Babel config given that if we did, it would be probably be 

- always use @babel/preset-env
- if you're using React, add @babel/preset-react
- if you're using TypeScript, add @babel/preset-typescript
- if you're using Flow, add @babel/preset-flow.

> I'd also advocate for better / more prescribed Typescript configs as I spent a lot of extra hours making Typescript feel right with the babel output for library consumption.

Not sure what we'd be recommending, the only thing that is really important when using Preconstruct is having noEmit: true?

---

Re having different Babel configs for different build types(probably just ESM/CJS vs UMD), I'm certainly open to it, I'll probably look at it after #331 and #332

**Comment by daphnesmit** [comment]
I changed my rollup config (esm + cjs builds) with `preserveModules:true` option to preconstruct build with the preset-env, preset-react and preset-typescript settings. 
The main  package where I import the npm package now got 60kb bigger (parsed) and it doesnt treeshake anymore like it did. Its a nextjs app.

Here are some screenshots of before and after changing rollup to preconstruct:

With preconstruct:
<img width="614" alt="Screenshot 2021-12-22 at 16 39 03" src="https://user-images.githubusercontent.com/4718795/147117560-976a8e45-a3b1-4287-adca-2a37d100c6a0.png">

<img width="1912" alt="Screenshot 2021-12-22 at 16 38 00" src="https://user-images.githubusercontent.com/4718795/147117537-30fab000-76b1-4e29-af1f-3489ba74d074.png">


Without preconstruct / with rollup ( preserveModules: true ) (seperate js files, not 1 bundle):
<img width="1914" alt="Screenshot 2021-12-22 at 17 08 57" src="https://user-images.githubusercontent.com/4718795/147121783-8b66c0af-44d3-4170-a6a3-e22a3017a3fd.png">
<img width="1914" alt="Screenshot 2021-12-22 at 17 08 57" src="https://user-images.githubusercontent.com/4718795/147121794-fcf1eac6-783e-45c6-afc3-6d74327e4d61.png">

**Comment by Andarist** [comment]
You might be using some non-treeshakable patterns there and `preserveModules:true` combined with `sideEffect:false` can sometimes produce better results in such situations. However, It's hard to know exactly what is going on without inspecting a repro case

**Comment by daphnesmit** [comment]
Yeah its a large repo where we also customly annotate hocs with babel (like chakra does aswell). might have to dive in to making a substract of the repo with just 2 components or such to analyze behaviour.
